app/src/main/java/com/doubtless/doubtless/main/bottomNav/RetroBottomNav.kt

```diff
@@ -0,0 +1,127 @@
+
+
+    override fun onAttachedToWindow() {
+        super.onAttachedToWindow()
+
+        elements.addAll(findBottomNavElements())
+
+        // set initial selected state
+        if (currentSelectedIndex == null) { // not restored after process death.
+            currentSelectedIndex = BottomNavMenu.defaultSelectedIndex
+        }
+
+        onNewSelectedIndex()
+        notifyAllUnSelectedElements()
+
+        // set click listeners on each button
+        elements.forEachIndexed { idx, btn ->
+            (btn as View).setOnClickListener {
+                setSelectStateOfSelectedItemAndNotifyUnselectedForElse(idx, btn)
+            }
+        }
+    }
+
+    private fun onNewSelectedIndex() {
+        onSelectedItemChangedListener?.onNewSelectedIndex(currentSelectedIndex!!)
+
+        elements[currentSelectedIndex!!].onSelected()
+
+        (elements[currentSelectedIndex!!] as View).layoutParams =
+            LayoutParams(0, LayoutParams.WRAP_CONTENT, 1f)
+    }
+
+    private fun notifyAllUnSelectedElements() {
+        // trigger onUnselected callback for rest of the elements.
+        elements.forEach { allElement ->
+            if (allElement != elements[currentSelectedIndex!!]) {
+                allElement.onUnselected()
+                (allElement as View).layoutParams =
+                    LayoutParams(0, LayoutParams.WRAP_CONTENT, 0.5f)
+            }
+        }
+    }
+
+    private fun setSelectStateOfSelectedItemAndNotifyUnselectedForElse(
+        idx: Int,
+        clickedBtn: BottomIntractableElement
+    ) {
+        if (currentSelectedIndex != idx) {  // notify on only new index selection.
+            currentSelectedIndex = idx
+            onSelectedItemChangedListener?.onNewSelectedIndex(idx)
+
+            clickedBtn.onSelected()
+
+            (elements[currentSelectedIndex!!] as View).layoutParams =
+                LayoutParams(0, LayoutParams.WRAP_CONTENT, 1f)
+        }
+
+        currentSelectedIndex = idx
+
+        notifyAllUnSelectedElements()
+    }
+
+    private fun findBottomNavElements(): List<BottomIntractableElement> {
+        val elementsFound = mutableListOf<BottomIntractableElement>()
+
+        this.children.forEach {
+            if (it is BottomIntractableElement) {
+                elementsFound.add(it)
+            }
+        }
+
+        return elementsFound
+    }
+
+    // ---------------------------------
+    // ------- save and restore --------
+
+    override fun onSaveInstanceState(): Parcelable? {
+        val bundle = Bundle()
+
+        if (currentSelectedIndex != null)
+            bundle.putInt("curr_selected", currentSelectedIndex!!)
+
+        return super.onSaveInstanceState()
+    }
+
+    override fun onRestoreInstanceState(state: Parcelable?) {
+        super.onRestoreInstanceState(state)
+
+        if (state is Bundle) {
+            currentSelectedIndex = state.getInt("curr_selected")
+        }
```

Comment:
Nit pick - curr_selected key can be extracted as a constant in this view itself

Reply:
ok